pkg/distributor/distributor_test.go

```diff
@@ -806,70 +806,101 @@ func TestShardCountFor(t *testing.T) {
+			wantShards:  1,
+			wantErr:     true,
 		},
 		{
-			name:           "not enough data to be sharded, stream size (18b) + ingested rate (0b) < 20b",
-			stream:         &logproto.Stream{Entries: []logproto.Entry{{Line: "a"}}},
-			rate:           0,
-			desiredRate:    20, // in bytes
-			wantStreamSize: 18, // in bytes
-			wantShards:     1,
-			wantErr:        false,
+			name:        "not enough data to be sharded, stream size (18b) + ingested rate (0b) < 20b",
+			stream:      &logproto.Stream{Entries: []logproto.Entry{{Line: "a"}}},
+			rate:        0,
+			desiredRate: 20, // in bytes
+			pushSize:    18, // in bytes
+			wantShards:  1,
+			wantErr:     false,
 		},
 		{
-			name:           "enough data to have two shards, stream size (36b) + ingested rate (24b) > 40b",
-			stream:         &logproto.Stream{Entries: []logproto.Entry{{Line: "a"}, {Line: "b"}}},
-			rate:           24, // in bytes
-			desiredRate:    40, // in bytes
-			wantStreamSize: 36, // in bytes
-			wantShards:     2,
-			wantErr:        false,
+			name:        "enough data to have two shards, stream size (36b) + ingested rate (24b) > 40b",
+			stream:      &logproto.Stream{Entries: []logproto.Entry{{Line: "a"}, {Line: "b"}}},
+			rate:        24, // in bytes
+			desiredRate: 40, // in bytes
+			pushSize:    36, // in bytes
+			wantShards:  2,
+			wantErr:     false,
 		},
 		{
 			// although the ingested rate by an ingester is 0, the stream is big enough to be sharded.
-			name:           "enough data to have two shards, stream size (36b) + ingested rate (0b) > 22b",
-			stream:         &logproto.Stream{Entries: []logproto.Entry{{Line: "a"}, {Line: "b"}}},
-			rate:           0,  // in bytes
-			desiredRate:    22, // in bytes
-			wantStreamSize: 36, // in bytes
-			wantShards:     2,
-			wantErr:        false,
+			name:        "enough data to have two shards, stream size (36b) + ingested rate (0b) > 22b",
+			stream:      &logproto.Stream{Entries: []logproto.Entry{{Line: "a"}, {Line: "b"}}},
+			rate:        0,  // in bytes
+			desiredRate: 22, // in bytes
+			pushSize:    36, // in bytes
+			wantShards:  2,
+			wantErr:     false,
 		},
 		{
 			name: "a lot of shards, stream size (1mb) + ingested rate (300mb) > 3mb",
 			stream: &logproto.Stream{Entries: []logproto.Entry{
 				{Line: "a"}, {Line: "b"}, {Line: "c"}, {Line: "d"}, {Line: "e"},
 			}},
-			rate:           0,  // in bytes
-			desiredRate:    22, // in bytes
-			wantStreamSize: 90, // in bytes
-			wantShards:     5,
-			wantErr:        false,
+			rate:        0,  // in bytes
+			desiredRate: 22, // in bytes
+			pushSize:    90, // in bytes
+			wantShards:  5,
+			wantErr:     false,
+		},
+		{
+			name:        "take push rate into account. Only generate two shards even though this push is quite large",
+			stream:      &logproto.Stream{Entries: []logproto.Entry{{Line: "a"}, {Line: "b"}}},
+			rate:        24,        // in bytes
+			pushRate:    1.0 / 6.0, // 6 pushes per second
```

Comment:
this confused me a little, why 6 pushes per second isn't `6.0 / 1.0` instead? it sounds like the more I push, the lower the push rate; I'd expect the opposite

Reply:
The comment is backward. This represents 1 push every 6 seconds. Fixed